driver-core/src/main/com/mongodb/internal/connection/OperationContext.java

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.internal.connection;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+public class OperationContext {
+    private static final AtomicLong NEXT_ID = new AtomicLong(0);
+    private final long id;
+
+    public OperationContext() {
+        id = NEXT_ID.incrementAndGet();
+    }
+
+    public long getId() {
+        return id;
```

Comment:
Are these ever created but not used (for example, optimistically), and if so, is it worth making this lazy to give the user cleaner numbering?

Reply:
I can't think of a case where that would happen.  At a minimum we're going to do server selection and log a message containing this id.